compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs

```diff
@@ -1236,11 +1239,7 @@ impl<'a> Context<'a> {
                 Ok(AcirValue::Array(values))
             }
             _ => unreachable!("ICE: Expected an array or numeric but got {ssa_type:?}"),
-        };
-
-        // Increment the var_index in case of a nested array
-        *var_index = self.acir_context.add_var(*var_index, one)?;
```

Comment:
This had been moved to outside of the value match. This breaks when returning an array as the var index assumes we are working with a fully flat list of memory. I'm assuming it was done while experimenting with the nested array tests breakage @jfecher. I have switched it back and added the simple repro for nested dynamic arrays.